### PR TITLE
feat: add restart hooks

### PR DESCRIPTION
**Pull request by ivan-tymoshenko**

### CODE REVIEW

**Review by mcollina — CHANGES_REQUESTED** [review]
Could you add docs?

File: index.js [hunk 1]
```diff
@@ -13,10 +13,17 @@ async function restartable (factory, opts, fastify = defaultFastify) {
...
+  const preRestartHooks = []
+  const onRestartHooks = []
+
   async function restart (restartOptions) {
     const requestListeners = server.listeners('request')
     const clientErrorListeners = server.listeners('clientError')
 
+    await Promise.allSettled(
+      preRestartHooks.map(hook => hook(app, restartOptions))
+    )
```
Comment (mcollina) [inline index.js:25]: Generally it's better to call hooks synchronously.

File: index.js [hunk 2]
```diff
@@ -46,6 +53,8 @@ async function restartable (factory, opts, fastify = defaultFastify) {
     await closeApplication(app)
 
     app = newApp
+
+    onRestartHooks.forEach(hook => hook(app, restartOptions))
```
Comment (mcollina) [inline index.js:57]: why is this sync?

Comment (ivan-tymoshenko) [reply index.js]: I rewrote this code, but the idea is that the `restart` function shouldn't wait for `onRestartHooks` execution. Does it make sense?

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**